scripts/generate_can.py

```diff
@@ -1,279 +1,299 @@
+                    offset = signal['offset']
+
+                    # calculate min/max based on scale/offset and length
+                    sig_min = - (2**(length-1)-1) * scale + \
+                        offset if is_signed else offset
+                    sig_max = (2**(length-1)-1) * scale + \
+                        offset if is_signed else (2**length-2) * scale + offset
+
+                # else if only given min/max (no scale/offset)
+                elif 'min' in signal and 'max' in signal:
+                    sig_min = signal['min']
+                    sig_max = signal['max']
+
+                    # calculate scale/offset based on min/max and length
+                    scale = (sig_max - sig_min) / (2**length - 2)
+                    offset = sig_min + (2**(length-1)-1) * \
+                        scale if is_signed else sig_min
+
+                else:
+                    print('Missing fields for signal "' + signal_name +
+                        '" in message "' + message_name + '"!')
+                    print('Supply either signal scale/offset or signal min/max')
+                    print('Script terminated.')
+                    exit()
+
+                # add these missing fields to the yaml
+                signal['scale'] = scale
+                signal['offset'] = offset
+                signal['min'] = sig_min
+                signal['max'] = sig_max
+
+                signals.append(
+                    cantools.database.can.Signal(
+                        name=signal_name,
+                        start=startbit,
+                        length=signal['length'],
+                        byte_order='little_endian',
+                        is_signed=signal['is_signed'],
+                        scale=signal['scale'],
+                        offset=signal['offset'],
+                        minimum=signal['min'],
+                        maximum=signal['max'],
+                        unit=signal['unit'],
+                        choices=signal['values'] if 'values' in signal else None,
+                        comment=signal['comment'],
+                        receivers=message['receivers']
+                    )
+                )
+
+                startbit += signal['length']
+
+            messages.append(
+                cantools.database.can.Message(
+                    frame_id=message['id'],
+                    name=message_name,
+                    length=message['size'],
+                    signals=signals,
+                    comment=message['comment'],
+                    senders=message['senders'],
+                    cycle_time=(None if message['frequency']
+                                == 0 else 1/message['frequency']),
+                )
+            )
+            # change value in "id:" section 
+            message['id'] = CAN_ID(message['id'])
+    
+    # converts roboteq CAN IDs to Hex
+    if can_yaml[bus_name]['roboteq_canids']:

```

Comment:
is this just because @cindyli-13 added custom handling for the roboteq canopen ids?

Reply:
When pyyaml converts the yaml file into a python object, it turns it into a python dictionary. Since the roboteq CAN IDs are keys rather than values, I need to use a separate method to replace the keys as hex. The CAN IDs in the other messages are values with the key "id" so it's much simplier.